dev/Experimental/floorplan_ilp/mip_ex.py

```diff
@@ -0,0 +1,481 @@
+                below = m.add_var(name=f'{a}_below_{b}', var_type=BINARY)
+                above = m.add_var(name=f'{a}_above_{b}', var_type=BINARY)
+                left_of = m.add_var(name=f'{a}_left_of_{b}', var_type=BINARY)
+                right_of = m.add_var(name=f'{a}_right_of_{b}', var_type=BINARY)
+                m += below + above + left_of + right_of == 1
+
+                m += m.var_by_name(f'{a}_urx') <= m.var_by_name(f'{b}_llx') + Mx - Mx*left_of
+                m += m.var_by_name(f'{b}_urx') <= m.var_by_name(f'{a}_llx') + Mx - Mx*right_of
+
+                m += m.var_by_name(f'{a}_ury') <= m.var_by_name(f'{b}_lly') + My - My*below
+                m += m.var_by_name(f'{b}_ury') <= m.var_by_name(f'{a}_lly') + My - My*above
+
+    # Between rows
+    between_rows = [m.add_var(name=f'between_rows_{i}') for i in range(len(fp)-1)]
+
+    for i, row in enumerate(fp):
+        if i == 0:
+            for cell in row:
+                m += m.var_by_name(f'{cell}_ury') <= m.var_by_name('ury')
+        elif i == len(fp) - 1:
+            for cell in row:
+                m += m.var_by_name('lly') <= m.var_by_name(f'{cell}_lly')
+
+        if i < len(between_rows):
+            for cell in row:
+                m += between_rows[i] <= m.var_by_name(f'{cell}_lly')
+        if i > 0:
+            for cell in row:
+                m += m.var_by_name(f'{cell}_ury') <= between_rows[i-1]
+
+    if symmetrize:
+        line_of_symmetry = m.add_var(name='line_of_symmetry')
+
+        for row in fp:
+            if len(row) % 2 == 1:
+                cell = row[len(row)//2]
+                m += m.var_by_name(f'{cell}_llx') + m.var_by_name(f'{cell}_urx') == 2*line_of_symmetry
+
+            for i in range(len(row)//2):
+                cell_l = row[i]
+                cell_r = row[len(row)-1-i]
+                m += m.var_by_name(f'{cell_l}_llx') + m.var_by_name(f'{cell_r}_urx') == 2*line_of_symmetry
+
+                # flips must be different
+                m += m.var_by_name(f'{cell_l}_fX') + m.var_by_name(f'{cell_r}_fX') == 1
+
+
+def place(*, sizes, wires, place_on_grid, fp, symmetrize=False, order=True, constraints=None, objective=None):
+    m = Model(sense=MINIMIZE, solver_name=CBC)
+    # set to one to see more progress output with the solver
+    m.verbose = 0
+
+    for k, _ in sizes.items():
+        for tag in ['llx', 'lly', 'urx', 'ury']:
+            m.add_var(name=f'{k}_{tag}', lb=0)
+
+        for tag in ['X', 'Y']:
+            f = m.add_var(name=f'{k}_f{tag}', var_type=BINARY)
+
+        size = dict(zip("xy", sizes[k]))
+
+        for axis in "xy":
+            m += m.var_by_name(f'{k}_ll{axis}') + size[axis] == m.var_by_name(f'{k}_ur{axis}')
+
+        if k in place_on_grid:
+            for d in place_on_grid[k]:
+                axis = d['axis']
+                pitch = d['pitch']
+                f = m.var_by_name(f'{k}_f{axis.upper()}')
+
+                if True:  # Refactored version
```

Comment:
@stevenmburns , thank you for this prototype, very helpful. I added a refactored, terse version for place on grid. It defines an additional redundant variable when there is a single offset.